dask_ml/model_selection/_hyperband.py

```diff
@@ -0,0 +1,580 @@
+    Parameters
+    ----------
+    SHA : SuccessiveHalvingSearchCV
+
+    Returns
+    -------
+    params : dict
+        Dictionary to re-create a SuccessiveHalvingSearchCV without the
+        estimator or parameters
+
+    Example
+    -------
+    >>> from sklearn.linear_model import SGDClassifier
+    >>> model = SGDClassifier()
+    >>> params = {"alpha": np.logspace(-1, 1)}
+    >>> SHA = SuccessiveHalvingSearchCV(model, params, tol=0.1,
+    ...                                 patience=True, random_state=42)
+    >>> _get_SHA_params(SHA)
+    {'aggressiveness': 3,
+     'max_iter': 100,
+     'n_initial_iter': 9,
+     'n_initial_parameters': 10,
+     'patience': True,
+     'random_state': 42,
+     'scoring': None,
+     'test_size': None,
+     'tol': 0.1}
+
+    """
+    return {
+        k: v
+        for k, v in SHA.get_params().items()
+        if "estimator_" not in k and k != "parameters" and k != "estimator"
+    }
+
+
+def _hyperband_paper_alg(R, eta=3):
+    """
+    Algorithm 1 from the Hyperband paper [1]_.
+
+    References
+    ----------
+    1. "Hyperband: A novel bandit-based approach to hyperparameter
+       optimization", 2016 by L. Li, K. Jamieson, G. DeSalvo, A. Rostamizadeh,
+       and A. Talwalkar.  https://arxiv.org/abs/1603.06560
+    """
+    s_max = math.floor(math.log(R, eta))
+    B = (s_max + 1) * R
+    brackets = reversed(range(int(s_max + 1)))
+    hists = {}
+    for s in brackets:
+        n = int(math.ceil(B / R * eta ** s / (s + 1)))
+        r = R * eta ** -s
+        r = int(r)
+        T = set(range(n))
+        hist = {
+            "num_estimators": n,
+            "estimators": {n: 0 for n in range(n)},
+            "decisions": [],
+        }
+        for i in range(s + 1):
+            n_i = math.floor(n * eta ** -i)
+            r_i = np.round(r * eta ** i).astype(int)
+            L = {model: r_i for model in T}
+            hist["estimators"].update(L)
+            hist["decisions"] += [r_i]
+            to_keep = math.floor(n_i / eta)
+            T = {model for i, model in enumerate(T) if i < to_keep}
+        hists[s] = hist
+    info = [
+        {
+            "bracket": k,
+            "estimators": hist["num_estimators"],
+            "partial_fit_calls": sum(hist["estimators"].values()),
+            "decisions": {int(h) for h in hist["decisions"]},
+        }
+        for k, hist in hists.items()
+    ]
+    return info
+
+
+def _get_patience(patience, max_iter, aggressiveness):
+    if not isinstance(patience, bool) and patience < max(max_iter // aggressiveness, 1):
+        msg = (
+            "Careful. patience={}, but values of patience=True (or "
+            "patience>={}) are recommended.\n\n"
+            "The goal of `patience` is to stop training estimators that have "
+            "already converged *when few estimators remain*."
+            "Setting patience=True accomplishes this goal.\n\n Please "
+            "continue with caution or good reason"
+        )
```

Comment:
Reword this. Make it more clear that Hyperband is already an adaptive scheme, and layering patience on top of it doesn't do much.